addons/binding/org.openhab.binding.meterreader/ESH-INF/thing/thing-types.xml

```diff
@@ -0,0 +1,77 @@
+    xmlns:config-description="http://eclipse.org/smarthome/schemas/config-description/v1.0.0"
+    xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+    <thing-type id="meter">
+        <label>Meter device</label>
+        <description>The meter device to read the SML or IEC 62056-21 messages from</description>
+
+        <config-description>
+            <parameter name="port" type="text">
+                <label>Connected serial port</label>
+                <description>The device serial port (e.g. /dev/tty0 or COM1)</description>
+                <required>true</required>
+                <limitToOptions>false</limitToOptions>
+                <context>serial-port</context>
+            </parameter>
+            <parameter name="refresh" type="integer">
+                <label>Refresh Rate</label>
+                <description>Refresh rate in seconds</description>
+                <default>20</default>
+                <unitLabel>s</unitLabel>
+            </parameter>
+            <parameter name="baudrateChangeDelay" type="integer">
+                <label>Delay of baudrate change in ms</label>
+                <default>0</default>
+                <description>USB to serial converters often require a delay of up to 250ms after the ACK before changing baudrate</description>
+            </parameter>
+            <parameter name="baudrate" type="text">
+                <label>Baudrate</label>
+                <default>AUTO</default>
+                <description>The baudrate of the serial port. If set to 'AUTO', it is dependent on the selected mode. The default is 300 baud for modes A, B, and C and 2400 baud for mode D, and 9600 baud for SML.</description>
+                <limitToOptions>false</limitToOptions>
+            </parameter>
+            <parameter name="mode" type="text">
+                <label>The protocol mode to use</label>
+                <default>SML</default>
+                <description>Can be SML (PUSH mode), Mode A,B,C (PULL)or D (PUSH)</description>
+            </parameter>
+            <parameter name="conformity" type="text">
+                <label>Conform to specific standard semantics</label>
+                <default>NONE</default>
+                <description>Reserved to conform to special semantics specified in specific standards. EDL_FNN: Currently applies the energy direction to WATT channels (which are absolute values) (see https://www.vde.com/resource/blob/951000/252eb3cdf1c7f6cdea10847be399da0d/fnn-lastenheft-edl-1-0-2010-01-13-data.pdf)</description>
+                <limitToOptions>true</limitToOptions>
+            </parameter>
+        </config-description>
+
+    </thing-type>
+
+    <channel-type id="NumberChannel">
```

Comment:
Those two channel types do not look right.
Note that [the granularity of channel types should be on its semantic level](https://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channels), i.e. if the value provided has a different meaning, it should be of a different channel type. This is imho specifically applicable here as you will have different dimensions for different channels and also want to provide different labels/descriptions/formatters or "advanced" flags.

I understand that there are likely too many channel types to create them manually (one per OBIS number, right?), but it should be feasible to implement a `ChannelTypeProvider` that dynamically constructs this information upon retrieval of the OBIS data.

How many channels are there created for an average meter? Are all values actually useful or might it make sense to restrict the binding to a subset, maybe leaving the user the option to manually add more channels (as per your example in the README)?

Reply:
I didn't know the `ChannelTypeProvider` to dynamically create ChannelTypes. I'll take a look and implement this (after my vacation ;-)).

But one question: I need the `itemType` as a String which also contains the `Dimension` (e.g. "`Number:Energy`"). From the device I get the value and the unit and then convert it to a `QuantityType`. To now get the `itemType` I need to reverse get the Dimension's name which is actually the simple class name of the `javax.measure.Quantity`. I haven't found a utility to get that Dimension (only vise versa with `UnitUtils#parseDimension`.

Reply:
You can get the Dimension as a class this way:
```
UnitUtils.parseUnit("1 " + yourUnitString).getDimension()
```
I don't see an easy way to get the String for the dimension, though. You'll probably have to use `instanceof Energy` etc. checks to identify the dimensions that you might come across in this binding.

Reply:
This is currently a blind spot in the UoM implementation. The I18nProviderImpl holds a map which provides the other way, finding a suitable (locale dependent) unit for a given dimension. For the time being I refused to do a full dimension <-> unit lookup table. Maybe a `unit.getSystemUnit` to `javax.measure.quantity.*` table would solve this.

> `UnitUtils.parseUnit("1 " + yourUnitString).getDimension()`

Be aware that the `javax.measure.Dimension` does not equal the dimension given for the item type. The item type extension always denotes an interface from `javax.measure.quantity`.

Reply:
So in short: There is no decent way to get the dimension as a string from a unit string 😠.
@msteigenberger The best suggestion for the moment is therefore that you define a mapping yourself, which provides the dimension string for a certain unit string, sorry...

Reply:
Yes, unfortunately this is the case right now.
Providing as custom mapping here makes sense as we don't need this mapping anywhere else (yet). Once other candidates pop up we might think of porting this to the framework.

Reply:
Hmmm... Why preserve a lookup table? You can always get the Quantity from a unit and from the Quantity implementation class you can get the  `javax.measure.quantity` interface via reflection API whose simple class name is the Dimension String in ESH. Am I overseeing something important?

Reply:
> You can always get the Quantity from a unit and from the Quantity implementation class you can get the javax.measure.quantity interface via reflection API whose simple class name is the Dimension String in ESH.

If this works, I'd agree that you can do without a lookup table.

Reply:
But we need that in ESH because the binding has no access to the `Quantity`. The easiest thing is to add a method in `QuantityType` to get the Dimension String. I could do a PR for that if it's fine for you (but only after vacation 😁).

Reply:
I let @htreu comment on whether he is fine with that approach and place :-)

Reply:
Yes, agreed, that would do. Go ahead @msteigenberger.

Reply:
I could not implement it the way I was thinking as the UOM implementation is not implementing the interfaces in `javax.measure.quantity` as I thought. I now made the assumption that all the Units are statically defined in a SystemOfUnits class (that's currently the case, namely `SIUnits`, `ImperialUnits`, `SmartHomeUnits` and `Units`). The specific `Quantity` dimension is a generic type parameter of the Unit. So I used all the statically defined Unit fields in the mentioned classes and compared the dimension with the dimension of the actual unit where we need the dimension string from. If its equal, we can use the simple name of the Quantity from the generic type parameter of the statically defined Unit.
https://github.com/eclipse/smarthome/pull/6303